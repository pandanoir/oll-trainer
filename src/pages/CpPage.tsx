import { useEffect, useState, VFC } from 'react';
import 'twin.macro';
import useTitle from 'react-use/lib/useTitle';
import { TopFace, Color, cpList } from '../data/cpList';
import { calculateScramble } from '../utils';
import {
  checkCpPattern,
  cpSwapPatterns,
  getCompanionSwap,
} from '../utils/checkCpPattern';
import { useInput } from '../utils/hooks/useInput';
import { useQuery } from '../utils/hooks/useQuery';

const Cube: VFC<{ color: string; onClick?: () => void }> = ({
  color,
  onClick,
}) => (
  <div
    onClick={onClick}
    style={{ width: 30, height: 30, background: color }}
  ></div>
);
const EmptyCube: VFC = () => (
  <div style={{ width: 30, height: 30, background: '#333' }}></div>
);
const CubeFace: VFC<{ cube: TopFace }> = ({ cube }) => (
  <div>
    {[0, 1, 2, 3, 4].map((row) => (
      <div
        key={row}
        style={{
          width: 150,
          display: 'flex',
          justifyContent: 'center',
        }}
      >
        {cube[row].map((color: Color | null, column: number) =>
          color === null ? (
            <EmptyCube key={column} />
          ) : (
            <Cube key={column} color={color} />
          )
        )}
      </div>
    ))}
  </div>
);

export const CpPage: VFC = () => {
  useTitle('Cp check');
  const query = useQuery();
  const { value, onChange } = useInput(query.get('solve') || '');
  const [topFaceType, setTopFaceType] = useState<null | string>(null);
  const [cpIndex, setCpIndex] = useState(0);
  useEffect(() => {
    try {
      const res = checkCpPattern(value);
      if (res) {
        setTopFaceType(res[0]);
        setCpIndex(res[1]);
      }
    } catch {
      setTopFaceType(null);
      setCpIndex(0);
    }
  }, [value]);
  return (
    <div>
      <h1>CP(Corner PLL)判断用</h1>
      <p>OLL の手順を入力するとそれに対応するCPを表示します。</p>
      <div tw="flex">
        <input
          type="text"
          value={value}
          tw="border rounded-md mt-1 mx-2 block flex-grow"
          onChange={onChange}
        />
      </div>
      {topFaceType && (
        <>
          逆手順: {calculateScramble(value)}
          <br />
          無交換(Skip,U,H,Z-perm): {topFaceType} {cpSwapPatterns[cpIndex]}
          <CubeFace cube={cpList[topFaceType][cpIndex]} />
          <br />
          対角交換(E,N,V,Y-perm): {topFaceType}{' '}
          {cpSwapPatterns[getCompanionSwap(cpIndex)]}
          <CubeFace cube={cpList[topFaceType][getCompanionSwap(cpIndex)]} />
          <br />
          隣接交換(A,F,G,J,R,T-perm): その他
        </>
      )}
    </div>
  );
};
