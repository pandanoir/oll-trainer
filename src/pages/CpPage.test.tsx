/**
 * @jest-environment jsdom
 */
import { fireEvent, render } from '@testing-library/react';
import { BrowserRouter } from 'react-router-dom';
import { CpPage } from './CpPage';

describe('CpPage', () => {
  it('snapshot', () => {
    const { asFragment, getByRole } = render(
      <BrowserRouter>
        <CpPage />
      </BrowserRouter>
    );
    expect(asFragment()).toMatchSnapshot();

    fireEvent.change(getByRole('textbox'), {
      target: { value: `F R U R' U' F'` },
    });

    expect(asFragment()).toMatchSnapshot();
  });
});
