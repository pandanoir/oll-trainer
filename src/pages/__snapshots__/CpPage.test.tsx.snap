// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Header snapshot 1`] = `
<DocumentFragment>
  <div>
    <h1>
      CP(Corner PLL)判断用
    </h1>
    <p>
      OLL の手順を入力するとそれに対応するCPを表示します。
    </p>
    <div
      class="_StyledDiv-sc-8uknjy-0 kzwVCR"
    >
      <input
        class="_StyledInput-sc-8uknjy-1 fLHGEY"
        type="text"
        value=""
      />
    </div>
  </div>
</DocumentFragment>
`;

exports[`Header snapshot 2`] = `
<DocumentFragment>
  <div>
    <h1>
      CP(Corner PLL)判断用
    </h1>
    <p>
      OLL の手順を入力するとそれに対応するCPを表示します。
    </p>
    <div
      class="_StyledDiv-sc-8uknjy-0 kzwVCR"
    >
      <input
        class="_StyledInput-sc-8uknjy-1 fLHGEY"
        type="text"
        value="F R U R' U' F'"
      />
    </div>
    逆手順: F U R U' R' F'
    <br />
    無交換(Skip,U,H,Z-perm): UCase diagonal
    <div>
      <div
        style="width: 150px; display: flex; justify-content: center;"
      >
        <div
          style="width: 30px; height: 30px; background: yellow;"
        />
        <div
          style="width: 30px; height: 30px; background: rgb(51, 51, 51);"
        />
        <div
          style="width: 30px; height: 30px; background: yellow;"
        />
      </div>
      <div
        style="width: 150px; display: flex; justify-content: center;"
      >
        <div
          style="width: 30px; height: 30px; background: green;"
        />
        <div
          style="width: 30px; height: 30px; background: orange;"
        />
        <div
          style="width: 30px; height: 30px; background: rgb(51, 51, 51);"
        />
        <div
          style="width: 30px; height: 30px; background: orange;"
        />
        <div
          style="width: 30px; height: 30px; background: blue;"
        />
      </div>
      <div
        style="width: 150px; display: flex; justify-content: center;"
      >
        <div
          style="width: 30px; height: 30px; background: rgb(51, 51, 51);"
        />
        <div
          style="width: 30px; height: 30px; background: rgb(51, 51, 51);"
        />
        <div
          style="width: 30px; height: 30px; background: yellow;"
        />
        <div
          style="width: 30px; height: 30px; background: rgb(51, 51, 51);"
        />
        <div
          style="width: 30px; height: 30px; background: rgb(51, 51, 51);"
        />
      </div>
      <div
        style="width: 150px; display: flex; justify-content: center;"
      >
        <div
          style="width: 30px; height: 30px; background: red;"
        />
        <div
          style="width: 30px; height: 30px; background: yellow;"
        />
        <div
          style="width: 30px; height: 30px; background: rgb(51, 51, 51);"
        />
        <div
          style="width: 30px; height: 30px; background: yellow;"
        />
        <div
          style="width: 30px; height: 30px; background: red;"
        />
      </div>
      <div
        style="width: 150px; display: flex; justify-content: center;"
      >
        <div
          style="width: 30px; height: 30px; background: green;"
        />
        <div
          style="width: 30px; height: 30px; background: rgb(51, 51, 51);"
        />
        <div
          style="width: 30px; height: 30px; background: blue;"
        />
      </div>
    </div>
    <br />
    対角交換(E,N,V,Y-perm): UCase no-swap
    <div>
      <div
        style="width: 150px; display: flex; justify-content: center;"
      >
        <div
          style="width: 30px; height: 30px; background: yellow;"
        />
        <div
          style="width: 30px; height: 30px; background: rgb(51, 51, 51);"
        />
        <div
          style="width: 30px; height: 30px; background: yellow;"
        />
      </div>
      <div
        style="width: 150px; display: flex; justify-content: center;"
      >
        <div
          style="width: 30px; height: 30px; background: green;"
        />
        <div
          style="width: 30px; height: 30px; background: orange;"
        />
        <div
          style="width: 30px; height: 30px; background: rgb(51, 51, 51);"
        />
        <div
          style="width: 30px; height: 30px; background: red;"
        />
        <div
          style="width: 30px; height: 30px; background: green;"
        />
      </div>
      <div
        style="width: 150px; display: flex; justify-content: center;"
      >
        <div
          style="width: 30px; height: 30px; background: rgb(51, 51, 51);"
        />
        <div
          style="width: 30px; height: 30px; background: rgb(51, 51, 51);"
        />
        <div
          style="width: 30px; height: 30px; background: yellow;"
        />
        <div
          style="width: 30px; height: 30px; background: rgb(51, 51, 51);"
        />
        <div
          style="width: 30px; height: 30px; background: rgb(51, 51, 51);"
        />
      </div>
      <div
        style="width: 150px; display: flex; justify-content: center;"
      >
        <div
          style="width: 30px; height: 30px; background: orange;"
        />
        <div
          style="width: 30px; height: 30px; background: yellow;"
        />
        <div
          style="width: 30px; height: 30px; background: rgb(51, 51, 51);"
        />
        <div
          style="width: 30px; height: 30px; background: yellow;"
        />
        <div
          style="width: 30px; height: 30px; background: red;"
        />
      </div>
      <div
        style="width: 150px; display: flex; justify-content: center;"
      >
        <div
          style="width: 30px; height: 30px; background: blue;"
        />
        <div
          style="width: 30px; height: 30px; background: rgb(51, 51, 51);"
        />
        <div
          style="width: 30px; height: 30px; background: blue;"
        />
      </div>
    </div>
    <br />
    隣接交換(A,F,G,J,R,T-perm): その他
  </div>
</DocumentFragment>
`;
